### DISCUSSION TITLE
add_spin=True (spin-polarized calculation) and adding SOC term

### DISCUSSION THREAD

**Discussion by dijasila** [TRIQS/solid_dmft#53] [Q&A]
Dear solid_dmft developers,
I have a few questions regarding spin-polarized calculations in particular, and some other questions in general. Please help me, I am new to DMFT. 

2. is it possible to do csc calculations with no spin polarization at DFT level and adding spinor=true at wannier and DMFT level?

3. for csc=False calculations, how can add soc term, is it added in seedname.inp file?

4. how can I specify the correct occupancy in seedname.inp file (the no. that goes on the 2nd line of seedname.inp)?, I have followed the other discussion thread but still is not clear to me. Does the number corresponds to the total number of electrons inside your wannier bands?

Please let me know if I have not made myself clear here.

**Comment by dijasila** [comment]
I have actually now done the dmft calculations by specifying,   lspinorb=.true. and     noncolin=.true., in seedname.scf.in and seedname.nscf.in files. and spinors=true in seedname.win file. 

But now I am facing the plotting issues, i.e. I get the dmft bands for spin=up and spin=down case fine, but at the same plot the tb_bands are plotted for both spin channels on the same graph. How can I sort this out?

**Reply by the-hampel** [reply]
Hi @dijasila, 

I will start with your first questions above. 

> how can I specify the correct occupancy in seedname.inp file (the no. that goes on the 2nd line of seedname.inp)?, I have followed the other discussion thread but still is not clear to me. Does the number corresponds to the total number of electrons inside your wannier bands?
yes you are supposed to specify the occupation that you expect to have in the energy window spanned by the Wannier bands. If you are unsure about this you can also let the converter determine the occupancy by switching once to the `bloch_mode=True` as converter options in the python call to the converter:
```
Converter = Wannier90Converter(seedname='seedname', bloch_basis=True, rot_mat_type='hloc_diag', add_lambda=None)
```
However, if you want to use the postprocessing routines of solid_dmft for plotting you need to switch this option back to `False` and use the resulting h5 archive. We can discuss how to get the spectral function with CSC and `bloch_mode=True` later. 

> is it possible to do csc calculations with no spin polarization at DFT level and adding spinor=true at wannier and DMFT level?

In principle this is possible yes. However, I believe as of now the calculations of the density correction is not yet checked / implemented for spin orbit coupling. @merkelm @phibeck Maybe you could give a more detailed answer here. 

> for csc=False calculations, how can add soc term, is it added in seedname.inp file?

You already tried and apparently it worked. Very good! I can only add that one should be careful with adding the SOC and spin on the DFT level. This is because you are then using the spin-polarized E_XC kernel which is very different from the non-spin polarized one. This could have an effect on how you choose interaction and double counting parameters. 

The other option for specific systems, i.e. t2g orbitals of a system with cubic symmetry, to add the SOC manually has an ad-hoc term in the converter by specifying `add_lambda` in the converter call. 

### For you latest question:

For plotting the results I believe there is no automatic way to separate down / up channels as soon as you use SOC. This is because you are coupling up / down spin channels. There is no way after applying SOC to know what is up or down without user guidance. If you are only performing a spin-polarized calculation you could separate up / down for plotting. As of now we do not have an option for this. If you need this let me know. It can be easily added. But again this will only work without SOC. 

However, you can use the option `proj_on_orb=[list of orbs to proj on]` and pass it to `get_dmft_bands`. This allows you to perform a projection on specific Wannier functions when calculating Akw. You can then choose to project on the Wannier functions that have a certain spin character. 

Best,
Alex

**Reply by dijasila** [reply]
Dear Alex, 
Many thanks for your reply, yes indeed the SOC mixes spin up and spin down channels, so it is not trivial to get up and down bands even at DFT level. 

however, for simple Spin-polarized calculations (with nspin=2), If we can add the option of plotting up and down,  DMFT and wannier90 bands, that will be great. If this can be done easily, I will be grateful, if you can develop a simple function for doing this. I could have tried to do it myself, but this would take me much more time, as I am less familiar with solid_dmft code, and would need to read many lines of codes before I can do this.

**Reply by the-hampel** [reply]
Actually I think it is possible with the function as it is right now. When calling `get_dmft_bands` pass `add_spin = False` and additionally specify in the sigma_dict passed to `get_dmft_bands`: 
```
sigma_dict_up = {....
                 'spin': 'up',   # choose here up or down. Depending on which spin channel you want to plot
                 ... other options}

sigma_dict_down = {...
                 'spin': 'down',   
                 ... other options}
```
now call the function twice: 
```
tb_bands, alatt_k_w_up, freq_dict = pcb.get_dmft_bands(..., **sigma_dict_down)
tb_bands, alatt_k_w_down, freq_dict = pcb.get_dmft_bands(..., **sigma_dict_down)
```

Then `alatt_k_w_up/down` holds the two spin channels separately. Does that work for you?

**Reply by dijasila** [reply]
Many thanks for your help, I did not realize you have already replied to my comments. 
that works now, apparently. 

I am facing another issue though, I will really appreciate if you could help me sort that out too. 
When I try to use a number bigger than 3 (no. of atoms) on the 3rd line of seedname.inp file, I get the following error in dmft.output....
why is it like that? It works fine as long as my no. is 3 or a smaller no. 

"File "/scratch/w27/sxa587/conda/lib/python3.10/site-packages/triqs/operators/util/U_matrix.py", line 578, in subarray
    return subarray(U, len(U.shape)*[(0,1,3)])
  File "/scratch/w27/sxa587/conda/lib/python3.10/site-packages/triqs/operators/util/U_matrix.py", line 578, in subarray
  File "/scratch/w27/sxa587/conda/lib/python3.10/site-packages/triqs/operators/util/U_matrix.py", line 578, in subarray
    return subarray(U, len(U.shape)*[(0,1,3)])
  File "/scratch/w27/sxa587/conda/lib/python3.10/site-packages/triqs/operators/util/U_matrix.py", line 578, in subarray
    sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3
    sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3
    sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3
    sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3
    sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3"

**Reply by the-hampel** [reply]
Hi, 
I am not sure. What is the number of orbitals of the atomic sites that you specified? Could it be that for this setup a Utensor cannot be constructed reasonably? How do you construct the Umatrix? 

Could you please upload the full output file. Then I can have a look. Thank you.

**Reply by dijasila** [reply]
0   4 4 3
 23.99996
 3
 0  0  2  5  0.00  0
 1  1  2  5  0.00  0
 2  2  2  5  0.00  0
For example, here is my seedname.inp file (that works). Although I have tried many different settings. The system that I am trying to study is MnSn in p62m symmetry. There are 6 atoms in the unit cell of the bulk MnSn. The d orbitals of the three Mn atoms is what I am trying to specify here. But in an alternative set of calculations I would like to include p orbitals of Sn atoms and hence would have to specify 6 atoms (on the 3rd line of seedname.inp
).

**Comment by dijasila** [comment]
I ideally i would like my dmft calculation to work for d orbitals of Mn , p orbitals of Mn and p orbitals of Sn. and seedname.inp would look like this.  (Note: I have tried many different settings i.e. by trying to specify the symmetrically equitant atoms using sort option, but did not get any reasonable results/ did not get it working)

 0   4 4 3
 53.99989
 6
 0  0  2  5  0.00  0
 1  1  2  5  0.00  0
 2  2  2  5  0.00  0
 0  0  1  3  0.00  0
 1  1  1  3  0.00  0
 2  2  1  3  0.00  0
 3  3  1  3  0.00  0
 4  4  1  3  0.00  0
 5  5  1  3  0.00  0

**Comment by the-hampel** [comment]
Hi @dijasila ,
I think the first number in each line specifying an atom needs to be a continuous index 0, 1, 2, 3, 4, 5, 6, ... . If you want to flag to atoms as equivalent use the second number (index) to mark them. However, the converter will complain if there are not found to be equivalent. Also the number in the third line should always be equivalent to the number of lines specified afterwards. I am not sure I understand why you have specified 9 lines. It should be 6 right?

**Reply by dijasila** [reply]
should not these be 9. 3 for d orbitals of Mn, 3 for p orbitals of Mn and 3 for p orbitals of Sn?

I have tried your suggestion, alas, I get the same error as described above. i.e. if I put any number above 3, I get "sa = a[tuple(slice(x) for x in a.shape[:n]) + (idxlist[n],)]
IndexError: index 3 is out of bounds for axis 3 with size 3""

**Reply by the-hampel** [reply]
Do you have 3 distinct sides of Mn? No right? Each lines specifies an impurity site that you want to treat with DMFT, which consists of multiple orbitals. This is an artificial helper construct. In principle Wannier90 does not care where the center of a WFs lives. However, we want to attach some physical meaning to it. So we tell the converter how many sites / impurities there are to be treated (third line of .inp file) first, and then specify one line for each site. I don't know how many sites you want to specifiy and how many orbitals there are at each site. 

Please show me a structure file, and the inp file so that I can judge if you wrote the inp file in the way you want it. You can just drag and drop files in this window and I will have a look then. Right now I am confused about how many Mn and Sn sites you have.

**Reply by Lanlin123** [reply]
Dear Alex,
I would like to ask a question about SOC here：
I have specified, lspinorb=.true. and noncolin=.true., in seedname.scf.in and seedname.nscf.in files. and spinors=true in seedname.win file, and SO=1 in seedname.inp and get the seedname.h5 file. But when I do the solid_dmft calculation, an error occurs:
 
![image](https://github.com/user-attachments/assets/147f49d6-b9ff-473d-9f74-eb4a26cf8297)
I don't find any parameters to control this in Solid_dmft. Can you give me some advice?

**Reply by the-hampel** [reply]
Hi @Lanlin123 , 
I think there is indeed a small bug in the code when the function `_adapt_U_4index_for_SO` is called. That is somewhere in line 568 (depending on the version). I just fixed this on the unstable branch of solid_dmft: 
https://github.com/TRIQS/solid_dmft/commit/d747ad52f999a16000cd578f71b6d33e5ff194cc

Which version are you using? Can you identify in your installation (`.../anaconda3/envs/triqs3.2/lib/python3.9/site-packages/solid_dmft/dmft_tools/interaction_hamiltonian.py`) that line as well, do the same change, and then try again? If the error persists please do open a full issue for this so we can better track it. If this solves it I will propagate the fix down to older versions. 

Best,
Alex

**Reply by Lanlin123** [reply]
Hi @the-hampel ,
I use version 3.2 and now it is fixed after modifying the code as you do. Thank you!

**Reply by the-hampel** [reply]
perfect. I pushed the fix also for the 3.3.x and 3.2.x branch.
